### PR TITLE
[deletions] new task abstraction and scheduler

### PR DESCRIPTION
**Pull request by dcramer**
This implements a new extensible layer for deletions allowing us to:

- Schedule deletions via a standard means (ScheduledDeletion)
- Abstraction to manage cascade deletes between both db models and other objects

It also includes:

- Compatibility shims so old tasks remain operable
- Change to Event deletions (less in-memory per task run)
- Moves several models to use bulk_delete_objects that are capable and previously werent

### CODE REVIEW

**Comment by dcramer** [comment]
I'm still working on getting tests in order. I will likely wait to transition legacy API calls until a followup patch.

**Comment by dcramer** [comment]
Need to figure out how I want to approach handling restore here yet. I might put the checks back in the individual delete_X tasks for legacy, and for future stuff make it actually revoke the ScheduleDeletion object.

**Comment by dcramer** [comment]
I think we're in a good spot for the initial feature now. There's a couple things that feel a bit odd (manager.get being used for subtasks, and how we optimize around batch deletions at a project level), but otherwise it works well.

**Review by mattrobenolt — APPROVED** [review]
wcgw

File: src/sentry/deletions/manager.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+__all__ = ['DeletionTaskManager']
+
+
+class DeletionTaskManager(object):
+    def __init__(self, default_task=None):
+        self.tasks = {}
+        self.default_task = default_task
+
+    def get(self, task=None, **kwargs):
```
Comment (dcramer) [inline src/sentry/deletions/manager.py]: this feels a little odd, and was needed since we cascade the manager into tasks (so its not using a singleton).

**Pull request merged**